include/circt/Dialect/Calyx/CalyxStructure.td

```diff
@@ -182,3 +162,67 @@ def CellOp : CalyxOp<"cell", [
   }];
   let verifier = "return ::verify$cppClass(*this);";
 }
+
+def GroupOp : CalyxOp<"group", [
+    HasParent<"WiresOp">,
+    NoRegionArguments,
+    RegionKindInterface,
+    SingleBlock,
+    Symbol
+  ]> {
+  let summary = "Calyx Group";
+  let description = [{
+    Represents a Calyx group, which is a collection
+    of assignments that are only active when the group
+    is run from the control execution schedule. A group
+    signifies its termination with a special port named
+    a "done" port.
+
+    ```mlir
+      calyx.group @MyGroup {
+        calyx.assign %1, %2 : i32
+        calyx.done %3 : i1
+      }
+    ```
+  }];
+
+  let arguments = (ins SymbolNameAttr: $sym_name);
+
+  let extraClassDeclaration = [{
+    // Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { return RegionKind::Graph; }
+  }];
+
+  let regions = (region SizedRegion<1>:$body);
+  let assemblyFormat = "$sym_name $body attr-dict";
+}
+
+// TODO(Calyx): Add optional guard to AssignOp and DoneOp.
+def AssignOp : CalyxOp<"assign", [
+    SameTypeOperands
+  ]> {
+  let summary = "Calyx Assignment";
+  let description = [{
+    The "calyx.assign" operation represents a non-blocking
+    assignment. An assignment may optionally be
+    guarded. This operation should only be instantiated in the
+    "calyx.wires" section or a "calyx.group".
+  }];
+  let arguments = (ins AnyType:$dest, AnyType:$src);
```

Comment:
For the record, starting with AnyType makes sense to me, and I'm not asking for any change or justification right now, just trying to provoke a little discussion :)

Reply:
As a little bit of background, Calyx currently has an extremely weak "width-only" type system for signals. That is, a 32-bit signed integer, unsigned integer, floating-point number, and pair of packed bfloats all have the same type: `32`. This has served our purposes pretty well for the most part, because the transformations we care about are on the whole not "type-sensitive" for lack of a better word. But we too have run into cases where a "real" type system for data values would be helpful, perhaps only as an option, but we haven't pursued that yet.

Reply:
Briefly looking at other type systems in CIRCT, it seems this would be similar to what `Comb` does here:
https://github.com/llvm/circt/blob/0797984c5d15fa9c8fc8070dcf39665d5aa8ffa9/include/circt/Dialect/Comb/Comb.td#L33-L38